### PR TITLE
[TT-XLA][PyTorch] DPO Training Workload for Gemma 1.1 2B Instruct - Math Preference

### LINKED ISSUE #225: ☂️ Training DPO workload for google/gemma-1.1-2b-it
Train google/gemma-1.1-2b-it using technique DPO

This training workload is necessary to showcase post-training alignment - same model as 26 but trained on a different task/dataset

### PR DESCRIPTION
**Pull request by mmilosevicTT**
### Issue
Closes https://github.com/tenstorrent/tt-blacksmith/issues/225

### Training Workload Setting

- **Model:** Gemma 1.1 2B Instruct
- **Strategy:** DPO on LoRA Model
- **Framework:** PyTorch
- **Dataset:** Math Preference
- **Hardware:** P150

### What's Changed
This PR introduces DPO training strategy on Gemma 1.1 2B Instruct on P150.

### Experiment Results
<table>
  <tr>
    <th>Training Accuracy</th>
    <th>Validation Accuracy</th>
  </tr>
  <tr>
    <td align="center">
W B Chart 22_06_2026, 14_14_42
    </td>
    <td align="center">
W B Chart 22_06_2026, 14_15_12
    </td>
  </tr>
  <tr>
    <th>Training Reward Margin</th>
    <th>Validation Reward Margin</th>
  </tr>
  <tr>
    <td align="center">
W B Chart 22_06_2026, 14_15_03
    </td>
    <td align="center">
W B Chart 22_06_2026, 14_15_25
    </td>
  </tr>
  <tr>
    <th>Training Loss</th>
    <th>Validation Loss</th>
  </tr>
  <tr>
    <td align="center">
W B Chart 22_06_2026, 14_14_55
    </td>
    <td align="center">
W B Chart 22_06_2026, 14_15_18
    </td>
  </tr>
</table>

### Checklist
- [x] Main training workflow
- [x] GPU testing
- [x] TT device testing
- [x] CI testing
    - [x] Existing tests are passing
    - [x] Register test to CI
    - [x] Upload golden files if necessary

### CODE REVIEW

File: blacksmith/datasets/torch/mathpreference/math_preference_dataset.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
+# SPDX-FileCopyrightText: (c) 2026 Tenstorrent AI ULC
+#
+# SPDX-License-Identifier: Apache-2.0
+"""
+Math Preference Dataset for DPO and SFT training.
```
Comment (umalesTT) [inline blacksmith/datasets/torch/mathpreference/math_preference_dataset.py]: I'd move this description to docstring of `MathPreferenceDataset` class, some things are duplicated this way and I think docstring is a better fit for description

File: blacksmith/datasets/torch/mathpreference/math_preference_dataset.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+        Returns both chosen and rejected sequences with labels. The SFT path
+        later derives its inputs from the chosen side.
+        """
+        instruction = example["instruction"]
+        chosen = example["chosen_response"]
+        rejected = example["rejected_response"]
+
+        prompt = self._format_prompt(instruction)
+
+        # Tokenize prompt to get its length
```
Comment (umalesTT) [inline blacksmith/datasets/torch/mathpreference/math_preference_dataset.py]: nit: Please add `.` at the end

File: blacksmith/datasets/torch/mathpreference/math_preference_utils.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
+# SPDX-FileCopyrightText: (c) 2026 Tenstorrent AI ULC
+#
+# SPDX-License-Identifier: Apache-2.0
+"""
```
Comment (umalesTT) [inline blacksmith/datasets/torch/mathpreference/math_preference_utils.py:4]: nit: I don't think reference to the paper should be here, I'd maybe even remove the whole docstring

File: blacksmith/experiments/torch/gemma11/dpo/train.py [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+from blacksmith.models.torch.huggingface.hf_models import get_model
+from blacksmith.tools.checkpoints_manager import CheckpointManager
+from blacksmith.tools.cli import generate_config, parse_cli_options
+from blacksmith.tools.device_manager import DeviceManager
+from blacksmith.tools.dpo_utils import compute_dpo_loss_from_batch
+from blacksmith.tools.logging_manager import TrainingLogger
+from blacksmith.tools.reproducibility_manager import ReproducibilityManager
+
+
+def _accumulate_metric_tensors(
```
Comment (umalesTT) [inline blacksmith/experiments/torch/gemma11/dpo/train.py]: This function, along with `_average_metric_tensors` should be moved to utils.py

File: blacksmith/experiments/torch/gemma11/dpo/train.py [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+        checkpoint_manager.load_checkpoint_path(config.sft_checkpoint_path, reference_model)
+    else:
+        logger.warning(
+            "No SFT checkpoint provided (sft_checkpoint_path is empty); "
+            "using the base pretrained model as the reference model. "
+            "For best results, first train an SFT model on chosen responses."
+        )
+    logger.info("Reference model loaded on device.")
+
+    for param in reference_model.parameters():
```
Comment (umalesTT) [inline blacksmith/experiments/torch/gemma11/dpo/train.py]: Is this needed? We're running reference model under `torch.no_grad` anyway, right?

File: blacksmith/experiments/torch/gemma11/dpo/train.py [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+
+    totals: dict[str, torch.Tensor | None] = {
+        "loss": None,
+        "accuracy": None,
+        "reward_margin": None,
+    }
+    num_batches = 0
+
+    # Use no_grad (not inference_mode): TT mark_argument_attributes reshapes inputs and
+    # fails with "Cannot set version_counter for inference tensor" under inference_mode.
```
Comment (umalesTT) [inline blacksmith/experiments/torch/gemma11/dpo/train.py]: Claude leftover? :)

File: blacksmith/experiments/torch/gemma11/dpo/train.py [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+    }
+    last_step_metrics = {}
+    last_val_metrics = {}
+
+    try:
+        if config.do_validation and val_dataloader is not None:
+            last_val_metrics = validate_dpo(
+                policy_model, reference_model, val_dataloader, config, device_manager, logger
+            )
+            logger.log_metrics({**last_val_metrics, "train/epoch": 0}, commit=True, step=global_step)
```
Comment (umalesTT) [inline blacksmith/experiments/torch/gemma11/dpo/train.py:204]: nit: "val" instead of "train/epoch"

File: blacksmith/tools/dpo_utils.py [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+from typing import Dict, Tuple
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+IGNORED_LABEL_ID = -100
+
+
+def get_batch_logps(
```
Comment (umalesTT) [inline blacksmith/tools/dpo_utils.py]: nit: maybe `get_batch_log_probs`?

**Review by umalesTT — APPROVED** [review]
LGTM!
Left a few, mainly cosmetic, comments

File: blacksmith/experiments/torch/gemma11/dpo/train.py (hunk 7 above)
Comment (mmilosevicTT) [reply blacksmith/experiments/torch/gemma11/dpo/train.py:204]: It might be confusing. It should be `train/epoch`, as we are logging info from before training. In the loop we log `train/epoch` when we log metrics to know the number of epoch model was trained for. I can remove this log if you want, but it is good to have 0 log meaning - no training done.

File: blacksmith/experiments/torch/gemma11/dpo/train.py (hunk 6 above)
Comment (mmilosevicTT) [reply blacksmith/experiments/torch/gemma11/dpo/train.py]: Removed

File: blacksmith/experiments/torch/gemma11/dpo/train.py (hunk 5 above)
Comment (mmilosevicTT) [reply blacksmith/experiments/torch/gemma11/dpo/train.py]: Not needed. Also we do not need to move to eval, as huggingface defaults to eval mode. Changed.

File: blacksmith/datasets/torch/mathpreference/math_preference_dataset.py (hunk 1 above)
Comment (mmilosevicTT) [reply blacksmith/datasets/torch/mathpreference/math_preference_dataset.py]: Removed this docstring and improved the class one.

File: blacksmith/datasets/torch/mathpreference/math_preference_utils.py (hunk 3 above)
Comment (mmilosevicTT) [reply blacksmith/datasets/torch/mathpreference/math_preference_utils.py:4]: Removed.

File: blacksmith/experiments/torch/gemma11/dpo/train.py (hunk 4 above)
Comment (mmilosevicTT) [reply blacksmith/experiments/torch/gemma11/dpo/train.py]: Moved 😄

File: blacksmith/tools/dpo_utils.py (hunk 8 above)
Comment (mmilosevicTT) [reply blacksmith/tools/dpo_utils.py]: Renamed.
